src/language/mongoDBService.ts

```diff
@@ -1,145 +1,162 @@
+}
 
 export default class MongoDBService {
+  _extensionPath?: string;
   _connection: Connection;
   _connectionId?: string;
   _connectionString?: string;
   _connectionOptions?: MongoClientOptions;
-  _cachedDatabases: CompletionItem[] | [] = [];
-  _cachedFields: { [namespace: string]: CompletionItem[] } | {} = {};
-  _cachedCollections: { [database: string]: CollectionItem[] } | {} = {};
-  _cachedShellSymbols: ShellCompletionItem;
-  _extensionPath?: string;
+
+  _cachedDatabases: CompletionItem[] = [];
+  _cachedFields: { [namespace: string]: CompletionItem[] } = {};
+  _cachedCollections: { [database: string]: CompletionItem[] } = {};
+  _cachedShellSymbols: { [symbol: string]: CompletionItem[] } = {};
+
   _visitor: Visitor;
+  _serviceProvider?: CliServiceProvider;
 
   constructor(connection: Connection) {
     this._connection = connection;
-    this._cachedShellSymbols = this._getShellCompletionItems();
     this._visitor = new Visitor(connection.console);
+
+    this._cacheShellSymbolsCompletionItems();
   }
 
-  // ------ CONNECTION ------ //
+  /**
+   * The connectionString used by LS to connect to MongoDB.
+   */
   get connectionString(): string | undefined {
     return this._connectionString;
   }
 
+  /**
+   * The connectionOptions used by LS to connect to MongoDB.
+   */
   get connectionOptions(): MongoClientOptions | undefined {
     return this._connectionOptions;
   }
 
+  /**
+   * The absolute file path of the directory containing the extension.
+   */
   setExtensionPath(extensionPath: string): void {
-    if (!extensionPath) {
-      this._connection.console.error(
-        'Set extensionPath error: extensionPath is undefined'
-      );
-    } else {
-      this._extensionPath = extensionPath;
-    }
+    this._extensionPath = extensionPath;
   }
 
-  async connectToServiceProvider(params: {
-    connectionId: string;
-    connectionString: string;
-    connectionOptions: MongoClientOptions;
-  }): Promise<boolean> {
-    this._clearCurrentSessionConnection();
-    this._clearCurrentSessionFields();
-    this._clearCurrentSessionDatabases();
-    this._clearCurrentSessionCollections();
-
-    this._connectionId = params.connectionId;
-    this._connectionString = params.connectionString;
-    this._connectionOptions = params.connectionOptions;
-
-    if (!this._connectionString) {
-      return Promise.resolve(false);
-    }
+  /**
+   * Connect to CliServiceProvider.
+   */
+  async connectToServiceProvider({
+    connectionId,
+    connectionString,
+    connectionOptions,
+  }: ServiceProviderParams): Promise<boolean> {
+    // If already connected close the previous connection.
+    await this.disconnectFromServiceProvider();
+
+    this._connectionId = connectionId;
+    this._connectionString = connectionString;
+    this._connectionOptions = connectionOptions;
+    this._serviceProvider = await CliServiceProvider.connect(
+      connectionString,
+      connectionOptions
+    );
```

Comment:
So, a question: This runs in the main thread, right, not inside a worker? Is there anything preventing us from using the `DataService` created by the `ConnectionController` for the operations for which we currently use the `CliServiceProvider` here? That would mean that we only need to connect once to a server, not twice, from the main thread, which seems reasonable to me

Reply:
I see … that should be fixable somehow since we don’t actually make use of the storage-mixin functionality :slightly_smiling_face: If it doesn’t work right now, the Browser-Compatible Data Service project should make this usable here. But yeah, definitely something for a separate PR in any case!

Reply:
And btw this is not a main thread, this is a LS separate thread already. I just tried a new DataService and it does not throw the error anymore there, we could replace CliServiceProvider with DataService in the mongoDBService.ts, but since it is a separate thread we can not just pass DataService, we need to pass the connection options and do a separate connect there. And I am thinking if we really want to pass two sets of connection options in a bit different format, one for CliServiceProvider and one for DataService 🤔

Reply:
Or we actually could pass only those that are meant for DataService and then do getMongoClientConnectionOptions.

Reply:
No, it’s not if it’s all on the same thread anyway :)